internal/provider/resource_offer.go

```diff
@@ -2,173 +2,255 @@ package provider
+	ModelName       types.String `tfsdk:"model"`
+	OfferName       types.String `tfsdk:"name"`
+	ApplicationName types.String `tfsdk:"application_name"`
+	EndpointName    types.String `tfsdk:"endpoint"`
+	URL             types.String `tfsdk:"url"`
+	// ID required by the testing framework
+	ID types.String `tfsdk:"id"`
+}
+
+func (o *offerResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_offer"
+}
+
+func (o *offerResource) Schema(_ context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = frameworkResSchema.Schema{
+		Description: "A resource that represent a Juju Offer.",
+		Attributes: map[string]frameworkResSchema.Attribute{
+			"model": frameworkResSchema.StringAttribute{
 				Description: "The name of the model to operate in.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"name": {
+			"name": frameworkResSchema.StringAttribute{
 				Description: "The name of the offer.",
-				Type:        schema.TypeString,
 				Optional:    true,
 				Computed:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"application_name": {
+			"application_name": frameworkResSchema.StringAttribute{
 				Description: "The name of the application.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"endpoint": {
+			"endpoint": frameworkResSchema.StringAttribute{
 				Description: "The endpoint name.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"url": {
+			"url": frameworkResSchema.StringAttribute{
 				Description: "The offer URL.",
-				Type:        schema.TypeString,
 				Computed:    true,
 			},
+			"id": frameworkResSchema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
 		},
 	}
 }
 
-func resourceOfferCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func (o *offerResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if o.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "offer", "create")
+		return
+	}
 
-	var diags diag.Diagnostics
-	modelName := d.Get("model").(string)
-	modelInfo, err := client.Models.GetModelByName(modelName)
-	if err != nil {
-		return diag.FromErr(err)
+	var data offerResourceModel
```

Comment:
s/data/plan. Didn't we decided that data from req.Plan would be `plan` and from req.State would be `state`?

Reply:
Yeah I'm a bit confused on that one https://github.com/juju/terraform-provider-juju/blob/0b67cc0f7a347bdabbb6c36c34b62178bfb83082/internal/provider/resource_machine.go#L217

I'll rename to "plan"/"state" 👍